---
title: "Null Distribution Histogram"
author: "Julia Craft"
date: "April 4, 2018"
output: html_document
---

**anything in all caps needs to be replaced with the appropriate data**

## Load libraries

```{r}
#install.packages("qqman")
library(ggplot2)
library(readr)
library(dplyr)
library(qqman)
```

## Parse data 

```{r}
NullEvalues <- read_tsv("outputRand2.txt")
Evalues <- read_tsv("outputCandida.txt")
```

## Plot data

```{r}
#Give column names to the null and relative data frames
colnames(Evalues) <- c("qseqid", "sseqid", "pident", "length", "mismatch", "gapopen","qstart", "qend", "sstart", "send", "evalue", "bitscore")
colnames(NullEvalues) <- c("qseqid", "sseqid", "pident", "length", "mismatch", "gapopen","qstart", "qend", "sstart", "send", "evalue", "bitscore")

#order the data frames from smallest to largest evalues
Evalues <- arrange(Evalues, evalue)
NullEvalues <- arrange(NullEvalues, evalue)

currentEvalues <- as.double(Evalues$evalue)
currentNullEvalues <- as.double(NullEvalues$evalue)

cutOffEvalue <- .1
currentNullEvalues <- currentNullEvalues[currentNullEvalues < cutOffEvalue]
currentEvalues <- currentEvalues[currentEvalues < cutOffEvalue]
nullPercentage = length(currentNullEvalues)/(length(currentNullEvalues) + length(currentEvalues))

allEvalues <- c(currentEvalues, currentNullEvalues)
allEvalues <- sort(allEvalues)

nullPercentage <- nrow(NullEvalues)/(nrow(NullEvalues) + nrow(Evalues))
iterator <- length(allEvalues)

smallest <- 1
smallestEvalue <- 100

while ((nullPercentage > .05) && (iterator >= 1)){
  iterator <- iterator - 1
  cutOffEvalue <- as.double(allEvalues[iterator])
  currentEvalues <- currentEvalues[currentEvalues < cutOffEvalue]
  currentNullEvalues <- currentNullEvalues[currentNullEvalues < cutOffEvalue]
  nullPercentage = (length(currentNullEvalues) + 1)/(length(currentNullEvalues) + length(currentEvalues) + 1)
  if(nullPercentage < smallest){
    smallest <- nullPercentage
    smallestEvalue <- cutOffEvalue
  }
}
smallest
smallestEvalue
```


```{r}

#Make a histogram with only the null distribution
cols <- c("Shuffled to Relative"="#f04546")
ggplot() +
    geom_histogram(data=NullEvalues, aes(NullEvalues$evalue, fill="Shuffled to Relative"), alpha = .3) +
    labs(title="Null Distribution", x="E-value", y="Count")+
    scale_colour_manual(values=cols) + scale_fill_manual(name="Alignments",values=cols)

#Make a histogram comparing the null and relative alignments with all returned data (E-value < 10)
cols <- c("Shuffled to Relative"="#f04546","Target to Relative"="#3591d1")
ggplot() + 
    geom_histogram(data=Evalues, aes(Evalues$evalue, fill="Target to Relative") , alpha = .3) +
    geom_histogram(data=NullEvalues, aes(NullEvalues$evalue, fill="Shuffled to Relative"), alpha = .3) +
    labs(title="Null Distribution and Results Distribution", x="E-value", y="Count")+
    scale_colour_manual(values=cols) + scale_fill_manual(name="Alignments",values=cols)
```

```{r}
#
nullLessThanOne <- select(NullEvalues, qseqid, sseqid, evalue)
nullLessThanOne <-nullLessThanOne[(nullLessThanOne[,3]< 0.1),]

relLessThanOne <- select(Evalues, qseqid, sseqid, evalue)
relLessThanOne <-relLessThanOne[(relLessThanOne[,3]< 0.1),]

pValue = .05/nrow(nullLessThanOne)
pValue

#Make a histogram with only the null distribution
cols <- c("Shuffled to Relative"="#f04546")
ggplot() +
    geom_histogram(data=nullLessThanOne, aes(nullLessThanOne$evalue, fill="Shuffled to Relative"), alpha = .3) +
    labs(title="Null Distribution of E-values Below 0.1", x="E-value", y="Count")+
    scale_colour_manual(values=cols) + scale_fill_manual(name="Alignments",values=cols) + 
    theme(axis.text=element_text(size=12), axis.title=element_text(size=14,face="bold")) + 
    geom_vline(aes(xintercept= smallestEvalue), coulour = "lightblue", linetype = "dashed")

#Make a histogram comparing the null and target alignments
cols <- c("Shuffled to Relative"="#f04546","Target to Relative"="#3591d1")
ggplot() + 
    geom_histogram(data=relLessThanOne, aes(relLessThanOne$evalue, fill="Target to Relative") , alpha = .3) +
    geom_histogram(data=nullLessThanOne, aes(nullLessThanOne$evalue, fill="Shuffled to Relative"), alpha = .3) +
    labs(title="Null Distribution and Results Distribution for E-values Below 0.1", x="E-value", y="Count")+
    scale_colour_manual(values=cols) + scale_fill_manual(name="Alignments",values=cols)
```

## Manhattan plots

```{r}
targetSeqs <- c()
relativeSeqs <- c()

for(i in 1:length(relLessThanOne$qseqid)){
  targetSeqs <- c(targetSeqs, as.integer(substr(relLessThanOne$qseqid[i], 8, 14)))
  relativeSeqs <- c(relativeSeqs, as.integer(substr(relLessThanOne$sseqid[i], 6, 10)))
}
relLessThanOne$target <- targetSeqs
relLessThanOne$relative <- relativeSeqs
relLessThanOne
manhattan(relLessThanOne,main = "PLACEHOLDER", ylim = c(0, 100) ,chr = "relative", bp = "target", p = "evalue", col = c("lightcoral", "lightblue"), ylab = "-log10(E-value)", xlab= "Relative Sequences",suggestiveline = F, genomewideline = F, chrlabs = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")) 

```


```{r}
positiveAlignments <- read_tsv("outputPositive.txt")
negativeAlignments <- read_tsv("outputNegative.txt")

#Give column names to the null and relative data frames
colnames(positiveAlignments) <- c("qseqid", "sseqid", "pident", "length", "mismatch", "gapopen","qstart", "qend", "sstart", "send", "evalue", "bitscore")
colnames(negativeAlignments) <- c("qseqid", "sseqid", "pident", "length", "mismatch", "gapopen","qstart", "qend", "sstart", "send", "evalue", "bitscore")

#positiveAlignments <- select(positiveAlignments, qseqid, sseqid, evalue)
#positiveAlignments <-positiveAlignments[(positiveAlignments[,3]< .1),]

#negativeAlignments <- select(negativeAlignments, qseqid, sseqid, evalue)
#negativeAlignments <-negativeAlignments[(negativeAlignments[,3]< .1),]


cols <- c("negative"="#f04546","positive"="#3591d1" )
ggplot() + 
    geom_histogram(data=positiveAlignments, aes(positiveAlignments$evalue, fill="positive") , alpha = .3) +
    geom_histogram(data=negativeAlignments, aes(negativeAlignments$evalue, fill="negative"), alpha = .3)+
    labs(title="Evalues for validation species alignment", x="E-value", y="Count")+
    scale_colour_manual(values=cols) + scale_fill_manual(name="Alignments",values=cols)

positiveAlignments$fromSpecies = c("positive")
negativeAlignments$fromSpecies = c("negative")
allResults <- rbind(positiveAlignments, negativeAlignments)


positiveSeqs <- c()
relativeSeqs <- c()

for(i in 1:length(allResults$qseqid)){
  relativeSeqs <- c(relativeSeqs, as.integer(substr(allResults$sseqid[i], 6, 10)))
  positiveSeqs <- c(positiveSeqs, as.integer(allResults$qseqid[i]))
}
allResults$qseqid <- positiveSeqs
allResults$relative <- relativeSeqs
allResults <- arrange(allResults, evalue)
allResults
manhattan(allResults,main = "E-values for Validation Alignments", ylim = c(-2, 10) ,chr = "relative", bp = "qseqid", p = "evalue", col = c("lightcoral"), ylab = "-log10(evalue)", xlab= "Relative Sequences", snp = "fromSpecies", highlight = c("positive")) 

```




